src/Microsoft.AspNet.Mvc.Core/ModelBinding/DictionaryModelBinder.cs

```diff
@@ -14,14 +17,86 @@ namespace Microsoft.AspNet.Mvc.ModelBinding
     public class DictionaryModelBinder<TKey, TValue> : CollectionModelBinder<KeyValuePair<TKey, TValue>>
     {
         /// <inheritdoc />
+        public override async Task<ModelBindingResult> BindModelAsync([NotNull] ModelBindingContext bindingContext)
+        {
+            var result = await base.BindModelAsync(bindingContext);
+            if (result == null || !result.IsModelSet)
+            {
+                // No match for the prefix at all.
+                return result;
+            }
+
+            var model = result.Model as Dictionary<TKey, TValue>;
+            Debug.Assert(model != null);
+            if (model.Count != 0)
+            {
+                // ICollection<KeyValuePair<TKey, TValue>> approach was successful.
+                return result;
+            }
+
+            var enumerableValueProvider = bindingContext.ValueProvider as IEnumerableValueProvider;
+            if (enumerableValueProvider == null)
+            {
+                // No IEnumerableValueProvider available for the fallback approach.
+                return result;
+            }
+
+            // Attempt to bind dictionary from a set of prefix[key]=value entries. Get the short and long keys first.
+            var keys = await enumerableValueProvider.GetKeysFromPrefixAsync(bindingContext.ModelName);
+            if (!keys.Any())
+            {
+                // No entries with the expected keys.
+                return result;
+            }
+
+            // Update the existing successful but empty ModelBindingResult.
+            var metadataProvider = bindingContext.OperationBindingContext.MetadataProvider;
+            var valueMetadata = metadataProvider.GetMetadataForType(typeof(TValue));
+            var valueBindingContext = ModelBindingContext.GetChildModelBindingContext(
+                bindingContext,
+                bindingContext.ModelName,
+                valueMetadata);
+
+            var modelBinder = bindingContext.OperationBindingContext.ModelBinder;
+            var validationNode = result.ValidationNode;
+
+            foreach (var key in keys)
+            {
+                var dictionaryKey = ConvertFromString(key.Key);
+                valueBindingContext.ModelName = key.Value;
+
+                var valueResult = await modelBinder.BindModelAsync(valueBindingContext);
+
+                // Always add an entry to the dictionary but validate only if binding was successful.
+                model[dictionaryKey] = ModelBindingHelper.CastOrDefault<TValue>(valueResult?.Model);
+                if (valueResult != null && valueResult.IsModelSet)
+                {
+                    validationNode.ChildNodes.Add(valueResult.ValidationNode);
```

Comment:
You mean for the extreme special case of a key type that implements `IValidatableObject`? Wasn't done in MVC 5 and I'm not sure it's worthwhile in MVC 6.

Further, in the collection of `KeyValuePair` case, MVC 5 simply skips `...Key` values that aren't instances of `TKey`. MVC 5 special-cases assignment of `null` to the key or value if either is a `ValueType` but MVC 6 does that at a lower level, in `TypeConverterModelBinder`.